pop-shell: [ERROR] mismatch on grab op entity

**Distribution (run `cat /etc/os-release`):**

```
➜ cat /etc/os-release
NAME="Pop!_OS"
VERSION="20.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 20.04 LTS"
VERSION_ID="20.04"
HOME_URL="https://system76.com/pop"
SUPPORT_URL="http://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
LOGO=distributor-logo-pop-os
```
 
**Related Application and/or Package Version (run `apt policy pop-shell` or provide the latest commit if building locally):**

```
➜ apt policy pop-shell
pop-shell:
  Installed: 0.1.0~1591631045~20.04~830dd7a
  Candidate: 0.1.0~1592948910~20.04~b739459
  Version table:
     0.1.0~1592948910~20.04~b739459 1001
       1001 http://ppa.launchpad.net/system76/pop/ubuntu focal/main amd64 Packages
       1001 http://ppa.launchpad.net/system76/pop/ubuntu focal/main i386 Packages
 *** 0.1.0~1591631045~20.04~830dd7a 100
        100 /var/lib/dpkg/status
```

**Issue/Bug Description:**

Some recent update introduced an issue where I will get desktop notifications `pop-shell: [ERROR] mismatch on grab op entity` randomly multiple times a day. This seems to occur when I am dragging windows or objects around on the desktop.  When the error pops up, the object I was dragging is dropped.    

**Steps to reproduce (if you know):**

Its pretty hard to reliably reproduce. Drag and snap windows. Move cards in Trello (chrome browser).

**Expected behavior:**

Don't drop my windows when I'm dragging them around.

**Other Notes:**

Please let me know what logs or steps you might need to troubleshoot.

Issue #398 had the same errors.

Here's some messages captured in syslog from the last time this happened.

```
Jul 28 11:00:17 mars gnome-shell[2970]: pop-shell: [ERROR] mismatch on grab op entity
Jul 28 11:00:17 mars gnome-shell[2970]: error: Pop Shell Error: mismatch on grab op entity
Jul 28 11:00:20 mars gnome-shell[2970]: Object .Gjs_ui_messageTray_Notification (0x55a17dd29a80), has been already deallocated — impossible to emit any signal on it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Jul 28 11:00:20 mars gnome-shell[2970]: == Stack trace for context 0x55a16da18900 ==
Jul 28 11:00:20 mars gnome-shell[2970]: #0   55a171ab3660 i   resource:///org/gnome/shell/ui/messageTray.js:515 (2ca4a2e78100 @ 127)
Jul 28 11:00:20 mars gnome-shell[2970]: #1   55a171ab35c0 i   resource:///org/gnome/shell/ui/messageTray.js:1513 (2ca4a2e7c970 @ 103)
Jul 28 11:00:20 mars gnome-shell[2970]: #2   55a171ab3538 i   resource:///org/gnome/shell/ui/messageTray.js:1497 (2ca4a2e7c8f8 @ 33)
Jul 28 11:00:20 mars gnome-shell[2970]: #3   55a171ab34a8 i   resource:///org/gnome/shell/ui/environment.js:75 (11a615c8d5b0 @ 98)
Jul 28 11:00:20 mars gnome-shell[2970]: #4   55a171ab3418 i   resource:///org/gnome/shell/ui/environment.js:149 (11a615c8d9e8 @ 14)
Jul 28 11:00:20 mars gnome-shell[2970]: == Stack trace for context 0x55a16da18900 ==
Jul 28 11:00:20 mars gnome-shell[2970]: #0   55a171ab3660 i   resource:///org/gnome/shell/ui/messageTray.js:516 (2ca4a2e78100 @ 142)
Jul 28 11:00:20 mars gnome-shell[2970]: #1   55a171ab35c0 i   resource:///org/gnome/shell/ui/messageTray.js:1513 (2ca4a2e7c970 @ 103)
Jul 28 11:00:20 mars gnome-shell[2970]: #2   55a171ab3538 i   resource:///org/gnome/shell/ui/messageTray.js:1497 (2ca4a2e7c8f8 @ 33)
Jul 28 11:00:20 mars gnome-shell[2970]: #3   55a171ab34a8 i   resource:///org/gnome/shell/ui/environment.js:75 (11a615c8d5b0 @ 98)
Jul 28 11:00:20 mars gnome-shell[2970]: #4   55a171ab3418 i   resource:///org/gnome/shell/ui/environment.js:149 (11a615c8d9e8 @ 14)
Jul 28 11:00:20 mars gnome-shell[2970]: Object .Gjs_ui_messageTray_Notification (0x55a17dd29a80), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Jul 28 11:00:20 mars gnome-shell[2970]: g_object_run_dispose: assertion 'G_IS_OBJECT (object)' failed
```

--- a/src/active_hint.ts
+++ b/src/active_hint.ts
@@ -8,43 +8,33 @@ import * as Ecs from 'ecs';
 
 const { GLib, St } = imports.gi;
 
-interface WindowDetails {
+interface Details {
+    sources: Array<number>;
+    workspace: number;
+}
+
+interface WindowDetails extends Details {
+    kind: 1;
     entity: Entity;
     meta: Meta.Window;
     parent: Clutter.Actor;
-    source1: number;
-    source2: number;
 }
 
 export class ActiveHint {
     dpi: number;
 
     private border: [Clutter.Actor, Clutter.Actor, Clutter.Actor, Clutter.Actor] = [
-        new St.BoxLayout({
-            reactive: false,
-            style_class: 'pop-shell-active-hint',
-            visible: false
-        }),
-        new St.BoxLayout({
-            reactive: false,
-            style_class: 'pop-shell-active-hint',
-            visible: false
-        }),
-        new St.BoxLayout({
-            reactive: false,
-            style_class: 'pop-shell-active-hint',
-            visible: false
-        }),
-        new St.BoxLayout({
-            reactive: false,
-            style_class: 'pop-shell-active-hint',
-            visible: false
-        })
+        new St.BoxLayout({ style_class: 'pop-shell-active-hint' }),
+        new St.BoxLayout({ style_class: 'pop-shell-active-hint' }),
+        new St.BoxLayout({ style_class: 'pop-shell-active-hint' }),
+        new St.BoxLayout({ style_class: 'pop-shell-active-hint' })
     ];
 
     private tracking: number | null = null;
 
-    window: WindowDetails | null = null;
+    tracked: WindowDetails | null = null;
+
+    restacker: SignalID = (global.display as GObject.Object).connect('restacked', () => this.restack_auto());
 
     constructor(dpi: number) {
         this.dpi = dpi;
@@ -63,7 +53,7 @@ export class ActiveHint {
     }
 
     is_tracking(entity: Entity): boolean {
-        return this.window ? Ecs.entity_eq(entity, this.window.entity) : false;
+        return this.tracked ? Ecs.entity_eq(entity, this.tracked.entity) : false;
     }
 
     position_changed(window: ShellWindow): void {
@@ -71,8 +61,38 @@ export class ActiveHint {
             this.hide();
         } else {
             this.show();
-            this.update_overlay();
+            this.update_overlay(window.rect());
         }
+    }
+
+    restack(actor: Clutter.Actor) {
+        if (this.tracked) {
+            if (this.tracked.workspace === global.workspace_manager.get_active_workspace_index()) {
+                // Avoid restacking if the window is maximized / fullscreen
+                if (this.tracked.kind === 1) {
+                    if (this.tracked.meta.get_maximized() !== 0 || this.tracked.meta.is_fullscreen()) return;
+                }
+
+                this.show();
+
+                for (const box of this.border) {
+                    global.window_group.set_child_above_sibling(box, actor);
+                }
+            } else {
+                this.hide();
+            }
+        }
+    }
+
+    restack_auto() {
+        if (!this.tracked) return;
+        let actor: null | Clutter.Actor = null;
+
+        if (this.tracked.kind === 1) {
+            actor = this.tracked.meta.get_compositor_private();
+        }
+
+        if (actor) this.restack(actor);
     }
 
     show() {
@@ -85,8 +105,8 @@ export class ActiveHint {
     track(window: ShellWindow) {
         this.disconnect_signals();
 
-        if (this.window) {
-            if (Ecs.entity_eq(this.window.entity, window.entity)) {
+        if (this.tracked) {
+            if (Ecs.entity_eq(this.tracked.entity, window.entity)) {
                 return;
             }
 
@@ -101,23 +121,29 @@ export class ActiveHint {
         const parent = actor.get_parent();
 
         if (parent) {
-            this.window = {
+            this.tracked = {
+                kind: 1,
                 entity: window.entity,
                 meta: window.meta,
                 parent: parent,
-                source1: window.meta.connect('size-changed', () => this.position_changed(window)),
-                source2: window.meta.connect('position-changed', () => this.position_changed(window)),
+                workspace: window.workspace_id(),
+                sources: [
+                    window.meta.connect('size-changed', () => this.position_changed(window)),
+                    window.meta.connect('position-changed', () => this.position_changed(window))
+                ],
             };
 
             this.tracking = GLib.idle_add(GLib.PRIORITY_LOW, () => {
                 this.tracking = null;
-                this.update_overlay();
+                this.update_overlay(window.rect());
 
                 this.show();
 
                 return false;
             });
         }
+
+        this.restack(actor);
     }
 
     untrack() {
@@ -125,48 +151,45 @@ export class ActiveHint {
 
         this.hide();
 
-        if (this.window) {
-            const actor = this.window.meta.get_compositor_private();
+        if (this.tracked) {
+            const actor = this.tracked.meta.get_compositor_private();
             if (actor) {
-                this.window.meta.disconnect(this.window.source1);
-                this.window.meta.disconnect(this.window.source2);
+                for (const s of this.tracked.sources.splice(0)) this.tracked.meta.disconnect(s);
             }
 
-            this.window = null;
+            this.tracked = null;
         }
     }
 
-    update_overlay() {
-        if (this.window) {
-            const rect = this.window.meta.get_frame_rect();
+    update_overlay(rect: Rectangular) {
+        const width = 3 * this.dpi;
 
-            const width = 3 * this.dpi;
+        const [w, n, e, s] = this.border;
 
-            const [left, top, right, bottom] = this.border;
+        w.x = rect.x - width;
+        w.y = rect.y;
+        w.width = width;
+        w.height = rect.height;
 
-            left.x = rect.x - width;
-            left.y = rect.y;
-            left.width = width;
-            left.height = rect.height;
+        e.x = rect.x + rect.width;
+        e.y = rect.y;
+        e.width = width;
+        e.height = rect.height;
 
-            right.x = rect.x + rect.width;
-            right.y = rect.y;
-            right.width = width;
-            right.height = rect.height;
+        n.x = rect.x - width;
+        n.y = rect.y - width;
+        n.width = (2 * width) + rect.width;
+        n.height = width;
 
-            top.x = rect.x - width;
-            top.y = rect.y - width;
-            top.width = (2 * width) + rect.width;
-            top.height = width;
-
-            bottom.x = rect.x - width;
-            bottom.y = rect.y + rect.height;
-            bottom.width = (2 * width) + rect.width;
-            bottom.height = width;
-        }
+        s.x = rect.x - width;
+        s.y = rect.y + rect.height;
+        s.width = (2 * width) + rect.width;
+        s.height = width;
     }
 
     destroy() {
+        global.display.disconnect(this.restacker);
+
         this.untrack();
 
         for (const box of this.border) {

--- a/src/auto_tiler.ts
+++ b/src/auto_tiler.ts
@@ -255,8 +255,8 @@ export class AutoTiler {
         }
     }
 
-    toggle_orientation(ext: Ext) {
-        const result = this.toggle_orientation_(ext);
+    toggle_orientation(ext: Ext, window: ShellWindow) {
+        const result = this.toggle_orientation_(ext, window);
         if (result.kind == ERR) {
             log.warn(`toggle_orientation: ${result.value}`);
         } else {
@@ -316,12 +316,7 @@ export class AutoTiler {
         log.info('\n\n' + buf);
     }
 
-    private toggle_orientation_(ext: Ext): Result<void, string> {
-        const focused = ext.focus_window();
-        if (!focused) {
-            return Err('no focused window to toggle');
-        }
-
+    private toggle_orientation_(ext: Ext, focused: ShellWindow): Result<void, string> {
         if (focused.meta.get_maximized()) {
             return Err('cannot toggle maximized window');
         }

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -222,6 +222,8 @@ export class Ext extends Ecs.System<ExtEvent> {
                         return;
                     }
 
+                    actor.remove_all_transitions();
+
                     event.window.meta.move_resize_frame(
                         true,
                         event.kind.rect.x,

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -460,9 +460,9 @@ export class Ext extends Ecs.System<ExtEvent> {
 
     on_active_workspace_changed() {
         const refocus_hint = () => {
-            if (!this.active_hint?.window) return
+            if (!this.active_hint?.tracked) return
 
-            let active = this.windows.get(this.active_hint.window.entity);
+            let active = this.windows.get(this.active_hint.tracked.entity);
             if (!active) return;
 
             let aws = this.workspace_id(active);
@@ -526,20 +526,15 @@ export class Ext extends Ecs.System<ExtEvent> {
     on_display_remove(id: number, display: Display) {
         if (!this.auto_tiler) return;
 
-        Log.info(`Display(${id}) removed`);
-
         let forest = this.auto_tiler.forest;
         let blocked = new Array();
 
         for (const [entity, [mon_id,]] of forest.toplevel.values()) {
-            Log.info(`Found TopLevel(${entity}, ${mon_id})`);
             if (mon_id === id) {
                 let fork = forest.forks.get(entity);
                 if (!fork) continue;
 
-                Log.info(`finding new workspace`);
                 const [new_work_id] = find_unused_workspace();
-                Log.info('finding monitor to retach');
                 const [new_mon_id, new_mon] = this.find_monitor_to_retach(display.area.width, display.area.height);
 
                 fork.workspace = new_work_id;
@@ -730,13 +725,17 @@ export class Ext extends Ecs.System<ExtEvent> {
             Log.error(`mismatch on grab op entity`);
         }
 
+        this.active_hint?.track(win);
+
         this.grab_op = null;
     }
 
     /** Triggered when a grab operation has been started */
     on_grab_start(meta: Meta.Window) {
         let win = this.get_window(meta);
-        if (win && win.is_tilable(this)) {
+        if (!win) return;
+
+        if (win.is_tilable(this)) {
             let entity = win.entity;
             let rect = win.rect();
 
@@ -746,6 +745,9 @@ export class Ext extends Ecs.System<ExtEvent> {
 
             this.size_signals_block(win);
         }
+
+
+        this.active_hint?.track(win);
     }
 
     on_gtk_shell_changed() {
@@ -852,12 +854,7 @@ export class Ext extends Ecs.System<ExtEvent> {
     }
 
     on_overview_hidden() {
-        if (this.active_hint && this.active_hint.window) {
-            let window = this.active_hint.window.meta;
-            if (!window.get_maximized()) {
-                this.active_hint.show();
-            }
-        }
+        this.active_hint?.restack_auto();
     }
 
     on_overview_shown() {
@@ -1009,6 +1006,8 @@ export class Ext extends Ecs.System<ExtEvent> {
                     this.register(Events.window_event(win, WindowEvent.Fullscreen));
                 }
             }
+
+            this.active_hint?.restack_auto();
         });
 
         this.connect(this.settings.ext, 'changed', (_s, key: string) => {

--- a/src/forest.ts
+++ b/src/forest.ts
@@ -208,7 +208,7 @@ export class Forest extends Ecs.World {
         workspace: number
     ): [Entity, Fork.Fork] {
         const entity = this.create_entity();
-        let orient = area && area.width > area.height ? Lib.Orientation.HORIZONTAL : Lib.Orientation.VERTICAL;
+        let orient = area.width > area.height ? Lib.Orientation.HORIZONTAL : Lib.Orientation.VERTICAL;
         let fork = new Fork.Fork(entity, left, right, area, workspace, orient);
 
         this.forks.insert(entity, fork);
@@ -245,43 +245,48 @@ export class Forest extends Ecs.World {
 
     /** Detaches an entity from the a fork, re-arranging the fork's tree as necessary */
     detach(fork_entity: Entity, window: Entity): [Entity, Fork.Fork] | null {
-        let reflow_fork = null;
+        const fork = this.forks.get(fork_entity);
+        if (!fork) return null;
 
-        this.forks.with(fork_entity, (fork) => {
+        let reflow_fork: [Entity, Fork.Fork] | null = null;
 
-            const parent = this.parents.get(fork_entity);
-            if (fork.left.is_window(window)) {
-                if (parent && fork.right) {
-                    reflow_fork = [parent, this.reassign_child_to_parent(fork_entity, parent, fork.right)];
-                } else if (fork.right) {
-                    reflow_fork = [fork_entity, fork];
-                    if (fork.right.kind == Node.NodeKind.WINDOW) {
-                        const detached = fork.right;
-                        fork.left = detached;
-                        fork.right = null;
-                    } else {
-                        this.reassign_children_to_parent(fork_entity, fork.right.entity, fork);
-                    }
+        const parent = this.parents.get(fork_entity);
+        if (fork.left.is_window(window)) {
+            if (parent && fork.right) {
+                const pfork = this.reassign_child_to_parent(fork_entity, parent, fork.right);
+                if (!pfork) return null;
+                reflow_fork = [parent, pfork];
+            } else if (fork.right) {
+                reflow_fork = [fork_entity, fork];
+                if (fork.right.kind == Node.NodeKind.WINDOW) {
+                    const detached = fork.right;
+                    fork.left = detached;
+                    fork.right = null;
                 } else {
-                    this.delete_entity(fork_entity);
+                    this.reassign_children_to_parent(fork_entity, fork.right.entity, fork);
                 }
-            } else if (fork.right && fork.right.is_window(window)) {
-                // Same as the `fork.left` branch.
-                if (parent) {
-                    reflow_fork = [parent, this.reassign_child_to_parent(fork_entity, parent, fork.left)];
-                } else {
-                    reflow_fork = [fork_entity, fork];
+            } else {
+                this.delete_entity(fork_entity);
+            }
+        } else if (fork.right && fork.right.is_window(window)) {
+            // Same as the `fork.left` branch.
+            if (parent) {
+                const pfork = this.reassign_child_to_parent(fork_entity, parent, fork.left);
+                if (!pfork) return null;
+                reflow_fork = [parent, pfork];
+            } else {
+                reflow_fork = [fork_entity, fork];
 
-                    if (fork.left.kind == Node.NodeKind.FORK) {
-                        this.reassign_children_to_parent(fork_entity, fork.left.entity, fork);
-                    } else {
-                        fork.right = null;
-                    }
+                if (fork.left.kind == Node.NodeKind.FORK) {
+                    this.reassign_children_to_parent(fork_entity, fork.left.entity, fork);
+                } else {
+                    fork.right = null;
                 }
             }
-        });
+        }
 
         if (reflow_fork) {
+            reflow_fork[1].rebalance_orientation();
         }
 
         return reflow_fork;

--- a/src/forest.ts
+++ b/src/forest.ts
@@ -75,7 +75,15 @@ export class Forest extends Ecs.World {
             const window = ext.windows.get(entity);
             if (!window) continue;
 
-            move_window(ext, window, r.rect, () => { });
+            let on_complete = () => { }
+            if (ext.tiler.window) {
+                if (Ecs.entity_eq(ext.tiler.window, entity)) {
+                    on_complete = () => {
+                        ext.set_overlay(window.rect());
+                    }
+                }
+            }
+            move_window(ext, window, r.rect, on_complete);
         }
 
         this.requested.clear();

--- a/src/fork.ts
+++ b/src/fork.ts
@@ -34,6 +34,9 @@ export class Fork {
     orientation_changed: boolean = false;
     is_toplevel: boolean = false;
 
+    /** Tracks toggle count so that we may swap branches when toggled twice */
+    private n_toggled: number = 0;
+
     constructor(entity: Entity, left: Node, right: Node | null, area: Rectangle, workspace: number, orient: Lib.Orientation) {
         this.area = area;
         this.left = left;
@@ -222,5 +225,15 @@ export class Fork {
             : Lib.Orientation.HORIZONTAL;
 
         this.orientation_changed = true;
+        if (this.n_toggled === 1) {
+            if (this.right) {
+                const tmp = this.right;
+                this.right = this.left;
+                this.left = tmp;
+            }
+            this.n_toggled = 0;
+        } else {
+            this.n_toggled += 1;
+        }
     }
 }

--- a/src/keybindings.ts
+++ b/src/keybindings.ts
@@ -25,7 +25,10 @@ export class Keybindings {
             "focus-down": () => ext.activate_window(ext.focus_selector.down(ext, null)),
             "focus-up": () => ext.activate_window(ext.focus_selector.up(ext, null)),
             "focus-right": () => ext.activate_window(ext.focus_selector.right(ext, null)),
-            "tile-orientation": () => ext.auto_tiler?.toggle_orientation(ext),
+            "tile-orientation": () => {
+                const win = ext.focus_window();
+                if (win) ext.auto_tiler?.toggle_orientation(ext, win);
+            },
             "toggle-floating": () => ext.auto_tiler?.toggle_floating(ext),
             "toggle-tiling": () => ext.toggle_tiling(),
         };

--- a/src/launcher.ts
+++ b/src/launcher.ts
@@ -157,7 +157,7 @@ export class Launcher extends search.Search {
             if (id >= this.selections.length) return;
 
             const selected = this.selections[id];
-            if (selected instanceof window.ShellWindow) {
+            if (selected && selected instanceof window.ShellWindow) {
                 if (selected.workspace_id() == ext.active_workspace()) {
                     const rect = selected.rect();
                     ext.overlay.x = rect.x
@@ -187,7 +187,6 @@ export class Launcher extends search.Search {
             }
 
             const launcher = MODES[this.mode].init(ext, this);
-            log.info(`Launcher Mode: "${launcher.prefix}"`);
             return launcher.apply(text.slice(launcher.prefix.length).trim(), index);
         };
 
@@ -210,7 +209,6 @@ export class Launcher extends search.Search {
                 for (const result of app_info.load_desktop_entries(path)) {
                     if (result.kind == OK) {
                         const value = result.value;
-                        log.info(value.display());
                         this.desktop_apps.push([where, value]);
                     } else {
                         const why = result.value;

--- a/src/log.ts
+++ b/src/log.ts
@@ -9,10 +9,7 @@ export function info(text: string) {
 }
 
 export function error(text: string) {
-    if (log_level > 1) {
-        log("[ERROR] " + text);
-        global.notify_error("Pop Shell Error", text);
-    };
+    if (log_level > 1) log("[ERROR] " + text);
 }
 
 export function warn(text: string) {

--- a/src/mod.d.ts
+++ b/src/mod.d.ts
@@ -25,7 +25,7 @@ declare interface GLib {
 
     spawn_command_line_sync(cmd: string): ProcessResult;
 
-    timeout_add(ms: number, priority: any, callback: () => Boolean): number;
+    timeout_add(priority: any, ms: number, callback: () => Boolean): number;
 }
 
 declare namespace GObject {

--- a/src/tiling.ts
+++ b/src/tiling.ts
@@ -29,7 +29,8 @@ enum Direction {
 export class Tiler {
     private keybindings: Object;
 
-    private window: Entity | null = null;
+    window: Entity | null = null;
+
     private swap_window: Entity | null = null;
 
     constructor(ext: Ext) {
@@ -38,8 +39,7 @@ export class Tiler {
                 if (ext.auto_tiler && this.window) {
                     const window = ext.windows.get(this.window);
                     if (window) {
-                        ext.auto_tiler.toggle_orientation(ext);
-                        ext.set_overlay(window.rect());
+                        ext.auto_tiler.toggle_orientation(ext, window);
                     }
                 }
             },

--- a/src/tiling.ts
+++ b/src/tiling.ts
@@ -3,7 +3,6 @@ const Me = imports.misc.extensionUtils.getCurrentExtension();
 
 import * as Lib from 'lib';
 import * as Tags from 'tags';
-import * as Log from 'log';
 import * as GrabOp from 'grab_op';
 import * as Rect from 'rectangle';
 import * as window from 'window';

--- a/src/tiling.ts
+++ b/src/tiling.ts
@@ -184,8 +184,6 @@ export class Tiler {
                 let fix_diff = () => {
                     let diff = before.diff(crect);
 
-                    Log.debug(`R DIFF BEFORE: ${diff.fmt()}`);
-
                     diff.width -= diff.width % 64;
                     diff.height -= diff.height % 64;
 
@@ -201,8 +199,6 @@ export class Tiler {
                         diff.height = (diff.height + 64) * -1;
                     }
 
-                    Log.debug(`R DIFF CORRECTED: ${diff.fmt()}`);
-
                     let tmp = before.clone();
                     tmp.apply(diff);
                     crect = tmp;
@@ -217,11 +213,8 @@ export class Tiler {
                     grab_op.rect = crect.clone();
                 };
 
-                Log.debug(`R BEFORE: ${crect.fmt()}`);
                 resize(mov1, callback);
-                Log.debug(`R MID: ${crect.fmt()}`);
                 resize(mov2, callback);
-                Log.debug(`R AFTER: ${crect.fmt()}`);
 
                 ext.auto_tiler.forest.arrange(ext, fork.workspace);
 
@@ -454,7 +447,6 @@ export class Tiler {
 
             if (!ext.auto_tiler || ext.contains_tag(win.entity, Tags.Floating)) {
                 // Make sure overlay is valid
-                global.log(`make sure overlay is valid`);
                 this.rect_by_active_area(ext, (_monitor, rect) => {
                     this.change(ext.overlay, rect, 0, 0, 0, 0);
                 });

--- a/src/tiling.ts
+++ b/src/tiling.ts
@@ -218,7 +218,7 @@ export class Tiler {
 
                 ext.auto_tiler.forest.arrange(ext, fork.workspace);
 
-                Tweener.on_tween_completion(window.meta, () => {
+                Tweener.on_window_tweened(window.meta, () => {
                     ext.register_fn(() => ext.set_overlay(window.rect()));
                 });
             }
@@ -309,7 +309,7 @@ export class Tiler {
         }
 
         if (watching) {
-            Tweener.on_tween_completion(watching.meta, () => {
+            Tweener.on_window_tweened(watching.meta, () => {
                 ext.register_fn(() => {
                     if (watching) {
                         ext.set_overlay(watching.rect());

--- a/src/tweener.ts
+++ b/src/tweener.ts
@@ -26,10 +26,20 @@ export function is_tweening(a: Clutter.Actor) {
         || a.get_transition('scale-x');
 }
 
-export function on_tween_completion(meta: Meta.Window, callback: () => void): SignalID {
-    return GLib.timeout_add(150, GLib.PRIORITY_DEFAULT, () => {
+export function on_window_tweened(meta: Meta.Window, callback: () => void): SignalID {
+    return GLib.timeout_add(GLib.PRIORITY_DEFAULT, 150, () => {
         const actor = meta.get_compositor_private();
         if (actor && is_tweening(actor)) return true;
+
+        callback();
+
+        return false;
+    });
+}
+
+export function on_actor_tweened(actor: Clutter.Actor, callback: () => void): SignalID {
+    return GLib.timeout_add(GLib.PRIORITY_DEFAULT, 150, () => {
+        if (is_tweening(actor)) return true;
 
         callback();
 

--- a/src/window.ts
+++ b/src/window.ts
@@ -204,7 +204,7 @@ export class ShellWindow {
                 }
 
                 ext.tween_signals.set(entity_string, [
-                    Tweener.on_tween_completion(this.meta, onComplete),
+                    Tweener.on_window_tweened(this.meta, onComplete),
                     onComplete
                 ]);
             } else {

--- a/src/window.ts
+++ b/src/window.ts
@@ -172,6 +172,9 @@ export class ShellWindow {
                 ext.register({ tag: 2, window: this, kind: { tag: 1, rect: clone } });
                 if (on_complete) ext.register_fn(on_complete);
                 ext.tween_signals.delete(entity_string);
+                if (ext.active_hint?.is_tracking(this.entity)) {
+                    ext.active_hint.show();
+                }
             };
 
             if (ext.animate_windows && !ext.init) {
@@ -195,6 +198,10 @@ export class ShellWindow {
                     duration: 149,
                     mode: null,
                 });
+
+                if (ext.active_hint?.is_tracking(this.entity)) {
+                    ext.active_hint.hide();
+                }
 
                 ext.tween_signals.set(entity_string, [
                     Tweener.on_tween_completion(this.meta, onComplete),

